hibernate-core/src/main/java/org/hibernate/event/internal/AbstractSaveEventListener.java

```diff
@@ -114,7 +114,7 @@ else if ( !generator.generatesOnInsert() ) {
 			// the @PrePersist callback to happen first
 			generatedId = null;
 		}
-		else if ( generatedBeforeExecution ) {
+		else if ( generatedBeforeExecution && ( ! generator.allowAssignedIdentifiers() || persister.getIdentifier( entity, source ) == null ) ) {
```

Comment:
While I agree with the scope of the change, I don't believe this is the correct way to do it. This way, when `allowAssignedIdentifier` returns `true` and there is a non-null value we do not call `generate` at all.

I would say we want to still invoke your custom `BeforeExecutionGenerator` though, and pass the existing identifier through the `currentValue` parameter.

Reply:
> I would say we want to still invoke your custom `BeforeExecutionGenerator` though, and pass the existing identifier through the `currentValue` parameter.

Yes, I agree. I think the bug is that `currentValue` is null when it shouldn't be.

Reply:
> I think the bug is that currentValue is null when it shouldn't be.

[That's what I was trying to say in the comment I left on the issue, though now I realize my comment wasn't very clear.]